### PR TITLE
✨ Add pyspellchecker

### PR DESCRIPTION
**Pull request by AminovE99**
:sparkles: Implemented pyspellchecke

Signed-off-by: Emil <amionv99@gmail.com>

### CODE REVIEW

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json [hunk 1]
```diff
@@ -121,7 +120,6 @@
    "bold": 0,
    "collapsible": 0,
    "columns": 0,
-   "depends_on": "eval:parent.file_to_read",
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json]: @AminovE99 Why has this been removed ?

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json [hunk 2]
```diff
@@ -187,7 +218,6 @@
...
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json]: @AminovE99 Why has this been removed ?

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json [hunk 3]
```diff
@@ -148,6 +146,39 @@
    "translatable": 0,
    "unique": 0
   },
+  {
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json]: @AminovE99 Add a `depends_on` `file_to_read`

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 4]
```diff
@@ -13,15 +13,35 @@
...
+    for word in misspelled:
+        corrected_word = spell_checker.correction(word)
+        print(word+ "instead of " + corrected_word)
+        message = message.replace(word, corrected_word)
+    return message
 
-        return text
+
+class OCRRead(Document):
+    def read_image(self, spell_checker):
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 Set `spell_checker` to `None` by default and do not execute if `None`

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 5]
```diff
@@ -13,15 +13,35 @@
...
+        message = read_document(self.file_to_read, self.language or 'en')
+        self.read_result = message
+        if self.spell_checker:
+            print("Spell checker has been enabled")
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 No `print`

File: requirements.txt [hunk 6]
```diff
@@ -3,4 +3,6 @@ six
 requests
 pillow
 wand
-tesserocr
\ No newline at end of file
+tesserocr
+spellchecker
+pyspellchecker
```
Comment (madmath03) [inline requirements.txt]: @AminovE99 Only keep one of them, do not need unused requirements.

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 7]
```diff
@@ -13,15 +13,35 @@
...
+    return word_list
+
+
+def analyze_text(message, lang='eng'):
+    spell_checker = SpellChecker()
+    only_words = get_words_from_text(message)
+    misspelled = spell_checker.unknown(only_words)
+    for word in misspelled:
+        corrected_word = spell_checker.correction(word)
+        print(word+ "instead of " + corrected_word)
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 No `print`

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 8]
```diff
@@ -13,15 +13,35 @@
...
 
-        self.read_result = text
-        self.save()
+def get_words_from_text(message):
+    message = message.replace(".", "").replace(",", "").replace("!", "").replace("\n", "")
+    word_list = list(filter(None, message.split()))
+    return word_list
+
+
+def analyze_text(message, lang='eng'):
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 Rename in `get_spellchecked_text` the function

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 9]
```diff
@@ -13,15 +13,35 @@
...
 
-class OCRRead(Document):
-    def read_image(self):
-        text = read_document(self.file_to_read, self.language or 'eng')
 
-        self.read_result = text
-        self.save()
+def get_words_from_text(message):
+    message = message.replace(".", "").replace(",", "").replace("!", "").replace("\n", "")
+    word_list = list(filter(None, message.split()))
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 Is it possible to use a regex `W` to automatically split on non-words ?

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py (hunk 4 above)
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 Also, do not forget to add a Unit Test with a Spell Checker!

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json [hunk 10]
```diff
@@ -55,7 +55,7 @@
    "bold": 0,
    "collapsible": 0,
    "columns": 0,
-   "depends_on": "",
+   "depends_on": "eval:parent.file_to_read",
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.json]: @AminovE99 You added a wrong dependency here : `file_to_read` cannot depend on `file_to_read`!

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 11]
```diff
@@ -13,15 +15,32 @@
...
+
+
+def get_spellchecked_text(message):
+    spell_checker = SpellChecker()
+    only_words = get_words_from_text(message)
+    misspelled = spell_checker.unknown(only_words)
+    for word in misspelled:
+        corrected_word = spell_checker.correction(word)
+        message = message.replace(word, corrected_word)
+    return message
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 The spellchecker does not have a language configuration ? If that's the case, we have a problem there.

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 12]
```diff
@@ -13,15 +15,32 @@
...
 
 class OCRRead(Document):
     def read_image(self):
-        text = read_document(self.file_to_read, self.language or 'eng')
-
-        self.read_result = text
+        message = read_document(self.file_to_read, self.language or 'en')
+        self.read_result = message
+        if self.spell_checker:
+            self.read_result = get_spellchecked_text(message, self.language or 'eng')
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 You removed the language parameter but still pass it.

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 13]
```diff
@@ -13,19 +15,40 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-        self.read_result = text
+    def read_image(self):
+        message = read_document(self.file_to_read, self.language or 'eng')
+        self.read_result = message
+        if self.spell_checker:
+            self.read_result = get_spellchecked_text(message, self.language)
         self.save()
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 If spell checker is enabled, it should replace the text read, not the text saved in the Doctype. Currently, the value returned by this function is different than the value saved in DB ! This is not what we want.

File: erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py [hunk 14]
```diff
@@ -13,19 +15,40 @@
...
-
-        return text
+        return message
 
 
 @frappe.whitelist()
-def read_document(path, lang='eng', event="ocr_progress_bar"):
+def read_document(path, lang, event="ocr_progress_bar"):
```
Comment (madmath03) [inline erpnext_ocr/erpnext_ocr/doctype/ocr_read/ocr_read.py]: @AminovE99 Why remove the default value of `lang` ??

File: erpnext_ocr/tests/test_tesseract.py [hunk 15]
```diff
@@ -16,14 +16,14 @@
 class TestTesseract(unittest.TestCase):
     def test_read_document_path_none(self):
         locale.setlocale(locale.LC_ALL, 'C')
-        recognized_text = read_document(None)
+        recognized_text = read_document(None, "eng")
```
Comment (madmath03) [inline erpnext_ocr/tests/test_tesseract.py]: @AminovE99 Just no...

File: erpnext_ocr/erpnext_ocr/doctype/ocr_language/ocr_language.py [hunk 16]
```diff
@@ -19,8 +19,7 @@ def check_language(lang):
 @frappe.whitelist()
 def lang_available(lang):
     """Call Tesseract OCR to verify language is available."""
-    if lang == 'en':
-        lang = "eng"
+    lang = "eng"
```
[the remark this answers is not in the document]
Comment (AminovE99) [reply erpnext_ocr/erpnext_ocr/doctype/ocr_language/ocr_language.py]: @madmath03 sorry. I don't know, that you will check it. I not done all of it

**Pull request merged**